### PR TITLE
mvcc: Clone for batch index compaction and shorten lock

### PR DESCRIPTION
**Pull request by jcalvert**
This is to address #9506 - When the BTree index grows to millions of entries, index compaction must iterate over all of them and this can take a substantial amount of time while continuing to hold the lock. This lock means that any concurrent read/writes will wait until compaction completes. By breaking compaction into batches of 10000, similar to the backend compaction, this allows relief for contention of the index lock. 

We came up with the following test as a way of validating that there is contention that blocks puts to the index while compaction is ongoing. On the `master` branch this test will fail, but it passes with our provided changes. We felt this test was useful for us in proving the issue but does not fit well into the existing test suite for etcd as far as we can tell. 

Coauthored with @cosgroveb

```go
func TestIndexCompactAndRuntime(t *testing.T) {                           
  ti := newTreeIndex()                                                    
  size := 1000000                                                         
  bytesN := 64                                                            
  keys := createBytesSlice(bytesN, size)                                  
  for i := 1; i < size-1; i++ {                                           
    ti.Put(keys[i], revision {main: int64(i), sub: int64(i)})             
  }                                                                       
  go ti.Compact(int64(500000))                                            
  time.Sleep(200000 * time.Nanosecond)                                  
  t1 := time.Now().UnixNano()                                             
  ti.Put(keys[size-1], revision {main: int64(size-1), sub: int64(size-1)})
  t2 := time.Now().UnixNano() - t1                                        
  if t2 > 150000000 {                                                     
    t.Errorf("Run time took too long! %v", t2)                            
  }                                                                       
}                                                                         
```

### CODE REVIEW

**Comment by xiang90** [comment]
The idea looks right. But as @heyitsanthony commented here: https://github.com/coreos/etcd/pull/9384#discussion_r172069649, we probably should have a better abstraction for traversing the tree in a mvcc manner.

**Comment by cosgroveb** [comment]
@xiang90 We're not sure that we fully understand the concerns being discussed in your link to #9384 and aren't super confident that we could be the ones to address the need for a better abstraction here without further guidance.

Are you looking for any particular changes from us or are you asking for us to wait until the problem being discussed in #9384 is settled? If you could provide us with a more concrete example or some psuedo-code we could certainly take a crack at it.

**Comment by jcalvert** [comment]
Based on the general store lock [here](https://github.com/coreos/etcd/blob/master/mvcc/kvstore.go#L214) this doesn't seem to really alleviate our problem fully and we are still investigating how to maintain consistent throughput during index compactions.

**Comment by xiang90** [comment]
@jcalvert 

Basically, we want to iterate the tree index without locking the lock for a long time. 

We can provide a generic func say: t.fuzzyAscend just like t.tree.Ascend to do it rather than exposing the details of locking in our business logic func.

**Comment by jcalvert** [comment]
@xiang90 

During more extensive testing, we discovered that the lock we previously mentioned was in fact preventing updates while the index was being compacted. Even batching in groups of ten thousand is enough to cause noticeable latency. We were finally able to achieve consistent throughput during compactions by pushing the lock on the tree index down into the `Ascend` function. In order to preserve traverse order, we used the `Clone` [function](https://godoc.org/github.com/google/btree#BTree.Clone) to produce a copy on write version of the tree so that we do not need to lock for the entire traversal. Please let us know any feedback you have, as we have experienced serious throughput degradation during index compactions.

**Comment by xiang90** [comment]
@jcalvert We cannot really modify the vendored btree. We have to push the change to upstream btree pkg first.

**Comment by xiang90** [comment]
> Even batching in groups of ten thousand is enough to cause noticeable latency. 

do you call yield to allow a reschedule of the go routine? i previously did some benchmark myself, and found batching of 10,000 should be good enough.

**Comment by jcalvert** [comment]
@xiang90 That change is already in a [more recent version of the btree package](https://github.com/google/btree/blob/master/btree.go#L612). Did I not do the vendoring process correctly?

**Comment by xiang90** [comment]
@jcalvert You probably need to first update the vendored pkg and the vendor lock file in another PR first.

@gyuho might help you on that.

File: mvcc/index_test.go [hunk 1]
```diff
@@ -17,7 +17,7 @@ package mvcc
 import (
 	"reflect"
 	"testing"
-
+	"time"
```
Comment (xiang90) [inline mvcc/index_test.go]: strange. why this is added?

**Comment by jcalvert** [comment]
> do you call yield to allow a reschedule of the go routine? i previously did some benchmark myself, and found batching of 10,000 should be good enough.

No, we did not try that. You mean adding `runtime.Gosched()`? That would be sufficient to allow others to acquire the lock?

**Comment by xiang90** [comment]
@jcalvert lazily clone is probably better if provided by the btree library already.

**Comment by xiang90** [comment]
@jcalvert rebase with current master?

File: mvcc/index_test.go (hunk 1 above)
Comment (jcalvert) [reply mvcc/index_test.go]: Not sure. Removing.

**Comment by jcalvert** [comment]
@xiang90 rebased. Thank you.

File: mvcc/index.go [hunk 2]
```diff
@@ -185,27 +184,32 @@ func (ti *treeIndex) RangeSince(key, end []byte, rev int64) []revision {
...
-			if ti.lg != nil {
-				ti.lg.Panic("failed to delete during compaction")
-			} else {
-				plog.Panic("store.index: unexpected delete failure during compaction")
+	clone := ti.tree.Clone()
+	ti.Unlock()
+
+	clone.Ascend(func(item btree.Item) bool {
+		keyi := item.(*keyIndex)
+		ti.Lock()
```
Comment (xiang90) [inline mvcc/index.go]: add comment on why the lock is needed?

File: mvcc/index_bench_test.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+	"testing"
+)
+
+func BenchmarkIndexCompact1(b *testing.B)   { benchmarkIndexCompact(b, 1) }
+func BenchmarkIndexCompact100(b *testing.B) { benchmarkIndexCompact(b, 100) }
+func BenchmarkIndexCompact10000(b *testing.B) { benchmarkIndexCompact(b, 10000) }
+func BenchmarkIndexCompact100000(b *testing.B) { benchmarkIndexCompact(b, 100000) }
+func BenchmarkIndexCompact1000000(b *testing.B) { benchmarkIndexCompact(b, 1000000) }
+
+func benchmarkIndexCompact(b *testing.B, size int) {
```
Comment (xiang90) [inline mvcc/index_bench_test.go]: do you have a result for the benchmark with a comparison with the result before the patch?

File: mvcc/kvstore.go [hunk 4]
```diff
@@ -245,9 +245,10 @@ func (s *store) Compact(rev int64) (<-chan struct{}, error) {
 	// ensure that desired compaction is persisted
 	s.b.ForceCommit()
 
-	keep := s.kvindex.Compact(rev)
 	ch := make(chan struct{})
 	var j = func(ctx context.Context) {
+		keep := s.kvindex.Compact(rev)
```
Comment (xiang90) [inline mvcc/kvstore.go]: why this needs to be changed?

File: mvcc/index.go [hunk 5]
```diff
@@ -17,7 +17,6 @@ package mvcc
 import (
 	"sort"
 	"sync"
-
```
Comment (xiang90) [inline mvcc/index.go]: revert this change.

File: mvcc/kvstore.go (hunk 4 above)
Comment (jcalvert) [reply mvcc/kvstore.go]: The defer of `s.mu.Unlock()` seems to mean that throughput still stalls during index compaction, although on audit of the code it isn't clear how this is the case. We can look to see if we can create a test to validate that.

Comment (xiang90) [reply mvcc/kvstore.go]: i mean why do we move the compact func into the schedule j unit?

File: mvcc/index_bench_test.go (hunk 3 above)
Comment (jcalvert) [reply mvcc/index_bench_test.go]: On local development machine -

master:
```
go test github.com/coreos/etcd/mvcc -v -run=^$ -bench BenchmarkIndexCompact
goos: linux
goarch: amd64
pkg: github.com/coreos/etcd/mvcc
BenchmarkIndexCompact1-16                1000000              1655 ns/op
BenchmarkIndexCompact100-16               100000             19772 ns/op
BenchmarkIndexCompact10000-16               2000            880883 ns/op
BenchmarkIndexCompact100000-16               200           8725995 ns/op
BenchmarkIndexCompact1000000-16              100         386072323 ns/op
PASS
```

this branch:
```
go test github.com/coreos/etcd/mvcc -v -run=^$ -bench BenchmarkIndexCompact  
goos: linux
goarch: amd64
pkg: github.com/coreos/etcd/mvcc
BenchmarkIndexCompact1-16                1000000              1572 ns/op
BenchmarkIndexCompact100-16               100000             22216 ns/op
BenchmarkIndexCompact10000-16               2000           1058001 ns/op
BenchmarkIndexCompact100000-16               100          10484775 ns/op
BenchmarkIndexCompact1000000-16              100         423129293 ns/op
```

Added time likely due to the additional lock/unlock calls.

File: mvcc/kvstore.go (hunk 4 above)
Comment (jcalvert) [reply mvcc/kvstore.go]: It seemed cleaner to put it there rather than remove the deferred unlock in place of explicitly unlocking in the error cases. This could be mistaken.

Comment (jcalvert) [reply mvcc/kvstore.go]: [Transactions](https://github.com/jcalvert/etcd/blob/48554b490a3fe57c6ae3f221f968f0835fb758ad/mvcc/kvstore_txn.go#L31) do indeed touch this lock so by offloading it to the scheduled action it means that transactions won't time out waiting for the lock. If you'd prefer us to change it so that the code explicitly unlocks we can do that.

File: mvcc/index.go (hunk 2 above)
Comment (jcalvert) [reply mvcc/index.go]: Added a comment about why the lock is needed. Let us know if this is not sufficient. Thank you.

File: mvcc/kvstore.go (hunk 4 above)
Comment (xiang90) [reply mvcc/kvstore.go]: i do not think this is safe since this func expects to be returned after the index is compacted (the old revisions are not reachable). we still need to keep the index compaction synchronously

File: mvcc/index_bench_test.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
+package mvcc
+
+import (
+	"testing"
+
+  "go.uber.org/zap"
```
Comment (xiang90) [inline mvcc/index_bench_test.go]: format? and why do we need this pkg?

**Comment by xiang90** [comment]
lgtm. defer to @gyuho

**Comment by gyuho** [comment]
@jcalvert Thanks again for hard work!

> copy on write version of the tree so that we do not need to lock for the entire traversal

Can you want to phrase this as a one-line release note here https://github.com/coreos/etcd/blob/master/CHANGELOG-3.4.md#improved, so people know who worked on this?

Also, do you have reproducible workloads that would benefit from this patch, so that I can cross-check on my side?

File: mvcc/index_bench_test.go [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+func BenchmarkIndexCompact1(b *testing.B)   { benchmarkIndexCompact(b, 1) }
+func BenchmarkIndexCompact100(b *testing.B) { benchmarkIndexCompact(b, 100) }
+func BenchmarkIndexCompact10000(b *testing.B) { benchmarkIndexCompact(b, 10000) }
+func BenchmarkIndexCompact100000(b *testing.B) { benchmarkIndexCompact(b, 100000) }
+func BenchmarkIndexCompact1000000(b *testing.B) { benchmarkIndexCompact(b, 1000000) }
+
+func benchmarkIndexCompact(b *testing.B, size int) {
+  log := zap.NewNop()
+  kvindex := newTreeIndex(log)
```
[the remark this answers is not in the document]
Comment (xiang90) [reply mvcc/index_bench_test.go]: hmm. why the tree index takes log as its arg? strange.

[the remark this answers is not in the document]
Comment (gyuho) [reply mvcc/index_bench_test.go]: @xiang90 I recently added structured logger as an option (where the logger object is created at top level and be passed downstream). Couldn't find an easier way to pass it around (will try to find a cleaner way). We can just pass `nil` for testing.

**Comment by gyuho** [comment]
@xiang90 Agree. I will find a cleaner way!

**Comment by jcalvert** [comment]
@gyuho 

`PUT` 10 million keys into server, send compaction request, immediately begin `PUT`, see timeouts. We discovered this trend by watching the timeouts from application servers in our Grafana dashboard. If you need, we can try to provide a client script to demonstrate. Thank you.

File: CHANGELOG-3.4.md [hunk 8]
```diff
@@ -34,6 +34,7 @@ See [code changes](https://github.com/coreos/etcd/compare/v3.3.0...v3.4.0) and [
...
 - Add [`snapshot`](https://github.com/coreos/etcd/pull/9118) package for easier snapshot workflow (see [`godoc.org/github.com/etcd/snapshot`](https://godoc.org/github.com/coreos/etcd/snapshot) for more).
 - Improve [functional tester](https://github.com/coreos/etcd/tree/master/functional) coverage: [proxy layer to run network fault tests in CI](https://github.com/coreos/etcd/pull/9081), [TLS is enabled both for server and client](https://github.com/coreos/etcd/pull/9534), [liveness mode](https://github.com/coreos/etcd/issues/9230), [shuffle test sequence](https://github.com/coreos/etcd/issues/9381), [membership reconfiguration failure cases](https://github.com/coreos/etcd/pull/9564), [disastrous quorum loss and snapshot recover from a seed member](https://github.com/coreos/etcd/pull/9565), [embedded etcd](https://github.com/coreos/etcd/pull/9572).
+- Improve [index compaction throughput](https://github.com/coreos/etcd/pull/9511) by using a copy on write clone to avoid holding the lock for the traversal of the entire index.
```
Comment (xiang90) [inline CHANGELOG-3.4.md]: it wont improve the compaction throughput. it solves the blocking problem.

Comment (jcalvert) [reply CHANGELOG-3.4.md]: 👍

**Comment by gyuho** [comment]
@jcalvert @xiang90  I resolved some formatting issue with license header. Otherwise looks good. Will merge after CI passes. Thanks a lot!

**Comment by xiang90** [comment]
@gyuho Can we squash the commits into one?

**Comment by gyuho** [comment]
@xiang90 Done.

**Comment by xiang90** [comment]
lgtm
